File: src/main/java/fr/umontpellier/iut/dominion/Player/BuySkills.java
package fr.umontpellier.iut.dominion.Player;

import fr.umontpellier.iut.dominion.*;
import fr.umontpellier.iut.dominion.cards.Card;
import fr.umontpellier.iut.dominion.cards.Events.Event;
import fr.umontpellier.iut.dominion.cards.component.TriggerComponent;

import java.util.ArrayList;
import java.util.List;

public class BuySkills implements PlayerComponent{
    private final Player self;
    public BuySkills(Player self) {
        this.self = self;
    }


    public void repayDebt(){
        int toRepay = Math.min(self.getValueOf(Item.MONEY), self.getValueOf(Item.DEBT));
        self.decrement(Item.DEBT, toRepay);
        self.decrement(Item.MONEY, toRepay);
    }

    public void useCoffre(){
        if(self.getValueOf(Item.COFFER) > 0){
            self.decrement(Item.COFFER, 1);
            self.increment(Item.MONEY, 1);
        }
    }

    public boolean canBuy(Card c) {
        boolean enoughMoney = self.getValueOf(Item.MONEY) >= c.getCost();
        boolean enoughPotion = self.getValueOf(Item.POTION) >= c.getPotion();
        boolean isNotDebted = self.getValueOf(Item.DEBT) == 0;
        boolean available = c.getAvailable().test(self);
        boolean isNotContraband = !self.getGame().getNamedCardsThisTurn("contraband").contains(c.getName());
        boolean isNotExpedition = (!c.hasType(CardType.EVENT) && !self.isFlagSet(Flags.expedition)) || c.hasType(CardType.EVENT);
        return enoughMoney && enoughPotion && isNotDebted && available && isNotContraband && isNotExpedition;
    }

    public int overPaid(Card c, int potion){
        int i = 0;
        if(c.hasType(CardType.OVERPAID) && self.getMoney() > c.getCost()){
            String choose = self.chooseWhatToDo("Do you want to overpaid ? ", List.of(c), Button.yesOrNo, true);
            if("y".equals(choose)) {
                int start = c.getCost();
                while(start < self.getMoney()){
                    List<Button> buttons = new ArrayList<>();
                    buttons.add(new Button("Increment", "i"));
                    buttons.add(new Button(" use Potion", "p"));

                    String howMuch = self.chooseWhatToDo("How much ? Max : " + self.getValueOf(Item.MONEY),  List.of(c), buttons , true);
                    if("i".equals(howMuch)) {
                        start += 1;
                    }else if("p".equals(howMuch)) {
                        potion += 1;
                    }else break;
                }
                i = start - c.getCost();
                c.set("OverpaidNumber", i);
                c.set("Potion", potion);
            }
        }
        c.getComponent(TriggerComponent.overPaidCard.class).ifPresent(t -> t.accept(self, c));
        return i;
    }

    public void buyCard(Card c) {
        if(c==null)return;
        int potion = 0;
        int overpaid = overPaid(c, potion);
        self.decrement(Item.MONEY, c.getCost() + overpaid);
        self.decrement(Item.POTION, c.getPotion() + potion);
        self.decrement(Item.DEBT, c.getDebt());
        self.log(self.toLog() + " bought " + c.toLog());
        Event event = new Event(c, Destination.DISCARD, self, true);
        if(c.hasType(CardType.EVENT)){
            if(c.canExecute(event, self, TriggerComponent.checkItSelfBuy.class))
                c.getComponent(TriggerComponent.checkItSelfBuy.class).ifPresent(t -> t.accept(event, c));
            return;
        }
        triggerBuy(event.getCard());
        self.gain(event);
        onCursePile(c);
    }

    public void triggerBuy(Card c){
        self.getCopyOf(Destination.INPLAY)
                .forEach(card -> card.getComponent(TriggerComponent.onBuy.class).ifPresent(d -> d.accept(self, c)));

    }

    public void onCursePile(Card c){
        if(self.getGame().hasToken(c.getName())){
            for (int i = 0; i <  self.getGame().getToken(c.getName()); i++ ){
                self.gain(self.getCardFromSupply("Curse"),Destination.DISCARD);
            }
        }
    }


}

File: src/main/java/fr/umontpellier/iut/dominion/Player/Skills/BuySkills.kt
package fr.umontpellier.iut.dominion.Player.Skills

import fr.umontpellier.iut.dominion.Annotation.AfterAllyTrigger
import fr.umontpellier.iut.dominion.Button
import fr.umontpellier.iut.dominion.CardType
import fr.umontpellier.iut.dominion.Enums.Locations.Destination
import fr.umontpellier.iut.dominion.Flags
import fr.umontpellier.iut.dominion.Item
import fr.umontpellier.iut.dominion.Player.Player
import fr.umontpellier.iut.dominion.Player.PlayerComponent.PlayerComponent
import fr.umontpellier.iut.dominion.Player.PlayerComponent.PlayerInTurnState
import fr.umontpellier.iut.dominion.cards.Card
import fr.umontpellier.iut.dominion.cards.Events.Event
import fr.umontpellier.iut.dominion.cards.component.TriggerComponent
import kotlinx.coroutines.flow.update
import org.hibernate.sql.ast.tree.expression.Over
import kotlin.math.min

class BuySkills : PlayerComponent {
}

fun Player.repayDebt(){
    val toRepay = min(getValueOf(Item.MONEY), getValueOf(Item.DEBT))
    decrement(Item.MONEY, toRepay)
    decrement(Item.DEBT, toRepay)
}


fun Player.useCoffer() = takeIf { getValueOf(Item.COFFER) > 0 }?.apply {
    decrement(Item.COFFER, 1)
    increment(Item.MONEY, 1)
}

fun Player.hasEnoughResourcesFor(card: Card): Boolean =
    getValueOf(Item.MONEY) >= card.costValue &&
            getValueOf(Item.POTION) >= card.potion &&
            getValueOf(Item.DEBT) == 0

fun Player.canBuy(card: Card): Boolean {
    if (!card.available(self)) return false
    if (!hasEnoughResourcesFor(card)) return false

    val isContraband = card.name in game.getNamedCardsThisTurn("contraband")
    val isDeludedAction = card.hasType(CardType.ACTION) && isFlagSet("Deluded")
    val isExpeditionRestricted = !card.hasType(CardType.EVENT) && getFlag(Flags.expedition).value

    return !isContraband && !isDeludedAction && !isExpeditionRestricted
}

private data class OverPaidResources(val cofferUsed: Int, val potionUsed : Int)

private suspend fun Player.overPaid(c: Card): OverPaidResources {
    var overpaidAmount = 0
    var overpaidPotion = 0

    val hasExtraMoney = getValueOf(Item.MONEY) > c.costValue
    val hasExtraPotion = getValueOf(Item.POTION) > c.potion

    if (c.hasType(CardType.OVERPAID) && (hasExtraMoney || hasExtraPotion)) {
        val choose = self.chooseWhatToDo("do you want to overpay?", listOf(c), buttons = Button.yesOrNo, canPass = true)

        if (choose == "y") {
            while (true) {
                val remainingMoney = getValueOf(Item.MONEY) - (c.costValue + overpaidAmount)
                val remainingPotion = getValueOf(Item.POTION) - (c.potion + overpaidPotion)

                val buttons = mutableListOf<Button>()
                if (remainingMoney > 0) { buttons.add(Button("Increment (+1 Coin)", "i")) }
                if (remainingPotion > 0) { buttons.add(Button("Use Potion (+1 Potion)", "p")) }

                if (buttons.isEmpty()) break

                val howMuch = self.chooseWhatToDo(
                    "overpaying: +$overpaidAmount Coins, +$overpaidPotion Potions. Choose what to add:",
                    listOf(c),
                    buttons = buttons,
                    canPass = true
                )

                when (howMuch) {
                    "i" -> overpaidAmount += 1
                    "p" -> overpaidPotion += 1
                    else -> break
                }
            }

            c.set("OverpaidNumber", overpaidAmount)
            c.set("Potion", overpaidPotion)
        }
    }

    c.getComponent<TriggerComponent.OverPaidCard>()?.let { it(self, c) }

    return OverPaidResources(overpaidAmount, overpaidPotion)
}

@AfterAllyTrigger
suspend fun Player.buyCard(c: Card?){
    if(c == null)return

    state.update { it.changeInTurnState(PlayerInTurnState.BuyingCard(c)) }

    suspend fun triggerBuy() {
        getCopyOf(Destination.PlayerZone.InPlay)?.forEach { card ->
            card.getComponent<TriggerComponent.OnBuy>()?.let { it(self, c) }
        }
    }


    suspend fun onCursePile() {
        if(game.hasToken(c.name))
            repeat(game.getToken(c.name)){
                gain(getCardFromSupply("Curse"), Destination.PlayerZone.Discard)
            }
    }

    val (overpaidAmount, overpaidPotion ) = overPaid(c)
    decrement(Item.MONEY, c.costValue + overpaidAmount)
    increment(Item.DEBT, c.debt)
    decrement(Item.POTION, c.potion + overpaidPotion)
    logBuy(c)

    val event = Event(c, Destination.PlayerZone.Discard, self, true)

    if(c.hasType(CardType.EVENT)){
        if (c.canExecute<TriggerComponent.CheckItSelfBuy>(event, self)) {
            c.getComponent<TriggerComponent.CheckItSelfBuy>()?.let { it(event, c) }
        }
        return
    }

    triggerBuy()
    gain(event)
    onCursePile()

}